sa/sa.go

```diff
@@ -456,22 +456,51 @@ func (ssa *SQLStorageAuthority) AddCertificate(ctx context.Context, req *sapb.Ad
 }
 
 // DeactivateRegistration deactivates a currently valid registration and removes its contact field
-func (ssa *SQLStorageAuthority) DeactivateRegistration(ctx context.Context, req *sapb.RegistrationID) (*emptypb.Empty, error) {
+func (ssa *SQLStorageAuthority) DeactivateRegistration(ctx context.Context, req *sapb.RegistrationID) (*corepb.Registration, error) {
 	if req == nil || req.Id == 0 {
 		return nil, errIncompleteRequest
 	}
-	_, err := ssa.dbMap.ExecContext(ctx,
-		"UPDATE registrations SET status = ?, contact = '[]' WHERE status = ? AND id = ? LIMIT 1",
-		string(core.StatusDeactivated),
-		string(core.StatusValid),
-		req.Id,
-	)
-	if err != nil {
-		return nil, err
+
+	result, overallError := db.WithTransaction(ctx, ssa.dbMap, func(tx db.Executor) (any, error) {
+		result, err := tx.ExecContext(ctx,
+			"UPDATE registrations SET status = ?, contact = '[]' WHERE status = ? AND id = ? LIMIT 1",
+			string(core.StatusDeactivated),
+			string(core.StatusValid),
+			req.Id,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("deactivating account %d: %w", req.Id, err)
+		}
+		rowsAffected, err := result.RowsAffected()
+		if err != nil || rowsAffected != 1 {
+			return nil, fmt.Errorf("deactivating account %d: no rows updated", req.Id)
```

Comment:
This error is not quite right; we could have two+ rows affected, or some other error. Probably the two rows thing is ignorable because we have a primary key. The "other error" case could be confusing if e.g. we got a connection reset in the middle?

Reply:
Hmm good point. The two-rows case is also largely ignorable because we have a "LIMIT 1" in the sql statement. And the other error case is largely ignorable (and ignored everywhere else we have code like this) because the Go sql driver [never actually returns an error there](https://cs.opensource.google/go/go/+/refs/tags/go1.24.1:src/database/sql/driver/driver.go;l=534), the err return is just to match an interface. So I'm not sure it's worth adding more messages here to handle those failure modes?

Reply:
Found some other code in the SA that handles each case individually and pulled that up here!